### PR TITLE
Merge ecflow workflow for running cycled parallel on WCOSS2

### PR DESCRIPTION
**Pull request by lgannoaa**
**Description**
Management requested to merge the ecflow workflow in preparation to run parallel with the dev-v16 branch on WCOSS2 that includes end to end jobs. 
#684

**Type of change**
- Include ecflow workflow files to run parallel on WCOSS2.
- Modify checkout and link script to checkout and install ecflow workflow for GFS cycled parallel on WCOSS2
- Fix config.gldas to allow uplevel assignment on CPCGAUGE
    Add WCOSS2 support to config.metp and config.vrfy
- Modified JGFS_ATMOS_CYCLONE_TRACKER to enable it for WCOSS2
    Add WCOSS2 support to verification and archive jobs arch.sh earc.sh metp.sh vrfy.sh
    Modify hpssarch_gen.sh to take log file from jobs
Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**
Currently is running cycled test on WCOSS2:
PSLOT: ecfops
HOMEgfs: /lfs/h2/emc/global/noscrub/Lin.Gan/para/packages/gfs.v16.2.0
COM: /lfs/h2/emc/ptmp/Lin.Gan/ecfops/com/prod/gfs/v16.2
log: /lfs/h2/emc/ptmp/Lin.Gan/ecfops/com/prod/gfs/v16.2/logs
EXPDIR: $HOMEgfs/parm/config
Verification Web site: https://www.emc.ncep.noaa.gov/gmb/Lin.Gan/metplus/ecfops

  
**Checklist**

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER [hunk 1]
```diff
@@ -138,6 +135,9 @@ elif [ $machine = LUNA -o $machine = SURGE -o $machine = WCOSS_C ] ; then # For
 elif [ $machine = VENUS -o $machine = MARS -o $machine = WCOSS_DELL_P3 ] ; then # For DELL
   machine=dell
   mpirun -n 1 ${USHens_tracker}/extrkr_gfs.sh ${loopnum} ${cmodel} ${CDATE} ${pert} ${DATA}
+elif [ $machine = WCOSS2 ] ; then # For WCOSS2
+  machine=wcoss2
+  mpirun -n 1 ${USHens_tracker}/extrkr_gfs.sh ${loopnum} ${cmodel} ${CDATE} ${pert} ${DATA}
```
Comment (aerorahul) [inline jobs/JGFS_ATMOS_CYCLONE_TRACKER]: I think the correct usage on wcoss2 is `mpiexec` and not `mpirun`.  Please confirm.

File: jobs/JGLOBAL_WAVE_PREP [hunk 2]
```diff
@@ -66,20 +66,20 @@ export COMIN=${COMIN:-${ROTDIR}/${CDUMP}.${PDY}/${cyc}/$COMPONENT}
...
-    $NLN $DMPDIR/${WAVECUR_DID}.${RPDY} $ROTDIR/${WAVECUR_DID}.${RPDY}
-  fi
-  if [ ! -L $ROTDIR/${CDUMP}.${PDY}/${cyc}/atmos/${WAVICEFILE} ]; then # Check if symlink already exists in ROTDIR
-    $NLN $DMPDIR/$CDUMP.${PDY}/$cyc/${WAVICEFILE} $ROTDIR/$CDUMP.${PDY}/$cyc/atmos/${WAVICEFILE}
-  fi
-  export COMIN_WAV_ICE=${COMIN_WAV_ICE:-$ROTDIR/$RUN.$PDY/$cyc/atmos}
-  export COMIN_WAV_RTOFS=${COMIN_WAV_RTOFS:-$ROTDIR}
-fi
+export COMIN_WAV_ICE=${COMIN_WAV_ICE:-$(compath.py ${envir}/obsproc/${obsproc_ver})}/${CDUMP}.${PDY}/${cyc}/atmos
+export COMIN_WAV_RTOFS=${COMIN_WAV_RTOFS:-$(compath.py ${envir}/${WAVECUR_DID}/${rtofs_ver})}
```
Comment (aerorahul) [inline jobs/JGLOBAL_WAVE_PREP]: The code on the left is for execution when NCO runs this.
Is this no longer NCO specific?
What about on R&D machines?

Does using the two lines above from 69, 70 remove the need for the commented code below on R&D machines?
Please confirm.

File: jobs/rocoto/arch.sh [hunk 3]
```diff
@@ -12,12 +12,6 @@
 ## cyc    : current cycle (HH)
 ###############################################################
 
-###############################################################
-# Source FV3GFS workflow modules
-. $HOMEgfs/ush/load_fv3gfs_modules.sh
```
Comment (aerorahul) [inline jobs/rocoto/arch.sh:17]: This will be required on the R&D machines to load HPSS module, no?
Can you provide a reason why this is deleted?

File: jobs/rocoto/earc.sh [hunk 4]
```diff
@@ -13,12 +13,6 @@
 ## ENSGRP : ensemble sub-group to archive (0, 1, 2, ...)
 ###############################################################
 
-###############################################################
-# Source FV3GFS workflow modules
-. $HOMEgfs/ush/load_fv3gfs_modules.sh
```
Comment (aerorahul) [inline jobs/rocoto/earc.sh:18]: same question as above.

File: jobs/rocoto/metp.sh [hunk 5]
```diff
@@ -13,14 +13,6 @@
 ## METPCASE : METplus verification use case (g2g1 | g2o1 | pcp1)
 ###############################################################
 
-###############################################################
-echo
-echo "=============== START TO SOURCE FV3GFS WORKFLOW MODULES ==============="
-. $HOMEgfs/ush/load_fv3gfs_modules.sh
```
Comment (aerorahul) [inline jobs/rocoto/metp.sh:19]: This might be non-consequential to remove, but good to verify on a R&D machine.

File: jobs/rocoto/vrfy.sh [hunk 6]
```diff
@@ -88,10 +80,16 @@ if [ $VRFYFITS = "YES" -a $CDUMP = $CDFNL -a $CDATE != $SDATE ]; then
     [[ ! -d $TMPDIR ]] && mkdir -p $TMPDIR
 
     xdate=$($NDATE -${VBACKUP_FITS} $CDATE)
-
-
     export RUN_ENVIR_SAVE=$RUN_ENVIR
     export RUN_ENVIR=$OUTPUT_FILE
+    export COMDAY=$PBS_O_WORKDIR
```
Comment (aerorahul) [inline jobs/rocoto/vrfy.sh]: $PBS_O_WORKDIR is a `PBS` specific variable and not defined on R&D machines w/ Slurm.
Please verify its compatibility on R&D machine.

File: parm/config/config.vrfy [hunk 7]
```diff
@@ -16,26 +16,24 @@ export CDFNL="gdas"               # Scores verification against GDAS/GFS analysi
...
+export VRFYPRCP="NO"              # Precip threat scores
+export VRFYRAD=${VRFYRAD:-"YES"}  # Radiance data assimilation monitoring
+export VRFYOZN=${VRFYOZN:-"YES"}  # Ozone data assimilation monitoring
+export VRFYMINMON=${VRFYMINMON:-"YES"} # GSI minimization monitoring
 export VRFYTRAK="YES"             # Hurricane track verification
 export VRFYGENESIS="YES"          # Cyclone genesis verification
-export VRFYFSU="NO"               # Cyclone genesis verification (FSU)
-export RUNMOS="NO"                # whether to run entire MOS package
+export VRFYFSU="YES"               # Cyclone genesis verification (FSU)
+export RUNMOS="YES"                # whether to run entire MOS package
```
Comment (aerorahul) [inline parm/config/config.vrfy]: Can you please explain why the defaults are changed?
Is it intended that `dev_v16` on R&D machines will run with these new defaults?

File: parm/config/config.vrfy [hunk 8]
```diff
@@ -16,26 +16,24 @@ export CDFNL="gdas"               # Scores verification against GDAS/GFS analysi
...
 
 #-------------------------------------------------
 # Fit to Observations
 #-------------------------------------------------
 
 if [ $VRFYFITS = "YES" ]; then
-
-    export fit_ver="newm.1.3"
-    export fitdir="$BASE_GIT/verif/global/Fit2Obs/${fit_ver}/batrun"
+    export fitdir="/lfs/h2/emc/global/noscrub/emc.global/git/Fit2Obs"
```
Comment (aerorahul) [inline parm/config/config.vrfy:36]: This is a hard-wired path to WCOSS2.  Is this appropriate or a temporary measure?

Comment (aerorahul) [reply parm/config/config.vrfy:36]: Seems it appropriate place would be in the WCOSS2 if-block, if necessary.

File: parm/config/config.vrfy [hunk 9]
```diff
@@ -142,22 +141,25 @@ fi
...
 if [ $machine = "WCOSS_DELL_P3" ] ; then
     export ens_tracker_ver=v1.1.15.3
 fi
 export HOMEens_tracker=$BASE_GIT/tracker/ens_tracker.${ens_tracker_ver}
 if [ $machine = "ORION" ] ; then
     export HOMEens_tracker=$BASE_GIT/tracker/TC_tracker.v1.1.15.2
 fi
-
+if [ $machine = "WCOSS2" ] ; then
+    export HOMEens_tracker=/lfs/h2/emc/global/noscrub/emc.global/git/ens_tracker.${ens_tracker_ver}
```
Comment (aerorahul) [inline parm/config/config.vrfy:153]: I believe @KateFriedman-NOAA will soon have an installation of the tracker in `BASE_GIT`.  Is there a reason to not have `$BASE_GIT` instead of the hard-wired path?

File: sorc/checkout.sh [hunk 10]
```diff
@@ -99,4 +99,13 @@ else
...
 fi
 
+echo wcoss2_ecf_cycled checkout ...
+if [[ ! -d wcoss2_ecf_cycled ]] ; then
+    rm -f ${topdir}/checkout-wcoss2_ecf_cycled.log
+    git clone --branch feature/wcoss2_ecf_cycled https://github.com/lgannoaa/ecflow.git wcoss2_ecf_cycled >> ${topdir}/checkout-wcoss2_ecf_cycled.log 2>&1
+    cd ${topdir}
+else
+    echo 'Skip. Directory wcoss2_ecf_cycled already exist.'
+fi
```
Comment (aerorahul) [inline sorc/checkout.sh]: What is this and why are these scripts to run ecflow not in the global-workflow repository?

More importantly, this is a private repository!!!

Screen Shot 2022-03-23 at 5 39 08 PM

File: sorc/link_fv3gfs.sh [hunk 11]
```diff
@@ -339,6 +339,15 @@ else
...
 
+#------------------------------
+#--link ecflow directories
+#------------------------------
+
+if [ $machine == "wcoss2" ]; then
+  cd $pwd/..
+  [[ -s ecf ]] && mv ecf ecf-old
+  $SLINK $pwd/wcoss2_ecf_cycled/ecf
+fi
```
Comment (aerorahul) [inline sorc/link_fv3gfs.sh]: In light of the above comment, this is not a valid solution.

File: ush/hpssarch_gen.sh [hunk 12]
```diff
@@ -234,12 +234,13 @@ if [ $type = "gdas" ]; then
...
-     echo  "./logs/${CDATE}/gdas${fstep}.log         " >>gdas.txt
-   fi
-  done
-  echo  "./logs/${CDATE}/gdaspost*.log               " >>gdas.txt
+#  for fstep in prep anal gldas fcst vrfy radmon minmon oznmon; do
+#   if [ -s $ROTDIR/logs/${CDATE}/gdas${fstep}.log ]; then
+#     echo  "./logs/${CDATE}/gdas${fstep}.log         " >>gdas.txt
+#   fi
+#  done
+#  echo  "./logs/${CDATE}/gdaspost*.log               " >>gdas.txt
```
Comment (aerorahul) [inline ush/hpssarch_gen.sh:242]: can you please explain what is the intent behind this?
Are logs not supposed to go to archive?

File: ush/hpssarch_gen.sh [hunk 13]
```diff
@@ -371,10 +372,10 @@ if [ $type = "enkfgdas" -o $type = "enkfgfs" ]; then
...
      fi 
   done # loop over FHR
-  for fstep in eobs eomg ecen esfc eupd efcs epos ; do
-   echo  "logs/${CDATE}/${CDUMP}${fstep}*.log        " >>enkf${CDUMP}.txt
-  done
-
+#  for fstep in eobs eomg ecen esfc eupd efcs epos ; do
+#   echo  "logs/${CDATE}/${CDUMP}${fstep}*.log        " >>enkf${CDUMP}.txt
+#  done
+  echo  "logs/${CDATE}/enkfgdas*.log        " >>enkf${CDUMP}.txt
```
Comment (aerorahul) [inline ush/hpssarch_gen.sh:378]: same question as above for `arch.sh`.

File: parm/config/config.metp [hunk 14]
```diff
@@ -14,56 +14,86 @@ export RUN_PRECIP_STEP1="YES"    # Run precip verification using METplus
...
-        export precip1_obtype="ccpa"
-        export precip1_accum_length="24"
-        export precip1_model_bucket_list="06"
-        export precip1_model_varname_list="APCP"
-        export precip1_model_fileformat_list="pgbf{lead?fmt=%H}.gfs.{init?fmt=%Y%m%d%H}.grib2"
-        export precip1_grid="G211"
-    fi
-fi
+## EMC_VERIF_GLOBAL SETTINGS
+export HOMEverif_global=${HOMEgfs}/sorc/verif-global.fd
```
Comment (aerorahul) [inline parm/config/config.metp:20]: I believe we might need SME to ensure these are correct for the `gdas` cycle, since the `gfs` cycle if-block has been removed.

I also note additional changes are made which might need checking due to the change in defaults in `config.vrfy`.

**Review by aerorahul — COMMENTED** [review]
@lgannoaa 
Thank you for your PR and changes.
I have a few comments that need clarifications.

While I understand this branch is currently in preparation for being used on the R&D machines, some of the edited/removed code might need to be added back.  To that end, we might need a different approach.

We can discuss that at a later time as the review progresses.

File: jobs/rocoto/metp.sh (hunk 5 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/metp.sh:19]: I know without even trying that MET+ uses modules. There are specific modules just for that in module_base.

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (WalterKolczynski-NOAA) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: Really these should all be using the launcher variable set by the ${machine}.env file and not have all this machine-branching code here at all.

File: jobs/rocoto/vrfy.sh (hunk 6 above)
Comment (WalterKolczynski-NOAA) [reply jobs/rocoto/vrfy.sh]: Should this not be $DATA?

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
I have pretty much the same questions as Rahul. Also, if code _is_ to be removed, it should just be removed, not commented out.

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (KateFriedman-NOAA) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: Yep, should be `mpiexec` and I second Walter's comment, we can set this in `$machine.env`.

File: jobs/JGLOBAL_WAVE_PREP (hunk 2 above)
Comment (KateFriedman-NOAA) [reply jobs/JGLOBAL_WAVE_PREP]: @JessicaMeixner-NOAA Should review this change, this is significant and needs to be vetted for use in ops since it will end up in the v16.3 upgrade. We went back and forth on this script a lot so changes must be considered/tested very carefully. Also, why is a block commented out but left in (lines 72-82)?

File: jobs/rocoto/arch.sh (hunk 3 above)
Comment (KateFriedman-NOAA) [reply jobs/rocoto/arch.sh:17]: I second Rahul's question, why is this deleted?

File: jobs/rocoto/vrfy.sh [hunk 15]
```diff
@@ -12,14 +12,6 @@
...
 ###############################################################
 
-###############################################################
-echo
-echo "=============== START TO SOURCE FV3GFS WORKFLOW MODULES ==============="
-. $HOMEgfs/ush/load_fv3gfs_modules.sh
-status=$?
-[[ $status -ne 0 ]] && exit $status
-
-
```
Comment (KateFriedman-NOAA) [inline jobs/rocoto/vrfy.sh:22]: Why is this removed?

File: parm/config/config.gldas [hunk 16]
```diff
@@ -10,7 +10,11 @@ echo "BEGIN: config.gldas"
 
 export GLDASSH=$HOMEgfs/scripts/exgdas_atmos_gldas.sh
 export gldas_spinup_hours=72
-export CPCGAUGE=$DMPDIR
+if [ $RUN_ENVIR = "emc" ]; then
+  export CPCGAUGE=${DMPDIR:-/lfs/h2/emc/global/noscrub/emc.global/dump}
+else
+  export CPCGAUGE=""
+fi
```
Comment (KateFriedman-NOAA) [inline parm/config/config.gldas]: Please explain this change. What is this fixing? What happens without this and what happens with this?

File: parm/config/config.metp (hunk 14 above)
Comment (KateFriedman-NOAA) [reply parm/config/config.metp:20]: @malloryprow should review this change since she built this config and updates it for us usually

File: parm/config/config.vrfy (hunk 8 above)
Comment (KateFriedman-NOAA) [reply parm/config/config.vrfy:36]: This should not be a hardcoded path...it should be `$BASE_GIT/Fit2Obs`. What happened to the `${fit_ver}/batrun` part of the path, is it no longer needed?

File: parm/config/config.vrfy [hunk 17]
```diff
@@ -52,6 +50,9 @@ if [ $VRFYFITS = "YES" ]; then
         export PREPQFITSH="$fitdir/subfits_cray_nems"
     elif [ $machine = "WCOSS_DELL_P3" ]; then
         export PREPQFITSH="$fitdir/subfits_dell_nems"
+    elif [ $machine = "WCOSS2" ]; then
+        export FITDIR=$fitdir
```
Comment (KateFriedman-NOAA) [inline parm/config/config.vrfy:54]: Why are there two fitdir paths? FITDIR and fitdir? Why not just use one?

File: parm/config/config.vrfy [hunk 18]
```diff
@@ -101,8 +102,6 @@ fi
 
 if [ $VRFYRAD = "YES" -o $VRFYMINMON = "YES" -o $VRFYOZN = "YES" ]; then
 
-    export envir="para"
```
Comment (KateFriedman-NOAA) [inline parm/config/config.vrfy:104]: Is this no longer needed?

File: parm/config/config.vrfy [hunk 19]
```diff
@@ -142,22 +141,25 @@ fi
 # Cyclone genesis and cyclone track verification
 #-------------------------------------------------
 
-export ens_tracker_ver=v1.1.15.1
+export ens_tracker_ver=v1.1.15.5
```
Comment (KateFriedman-NOAA) [inline parm/config/config.vrfy:144]: This should be set in a version file now (e.g. `versions/run.ver`).

File: parm/config/config.vrfy (hunk 9 above)
Comment (KateFriedman-NOAA) [reply parm/config/config.vrfy:153]: We should be trying to consolidate this whole section to just be 
`export HOMEens_tracker=$BASE_GIT/ens_tracker.${ens_tracker_ver}`
...where `ens_tracker_ver` can be set in `run.ver`. If the tracker version were to differ between machines then `ens_tracker_ver` would go into each `$target.ver` file.

File: sorc/checkout.sh (hunk 10 above)
Comment (KateFriedman-NOAA) [reply sorc/checkout.sh]: Strongly second @aerorahul 's question...what is this?

File: sorc/link_fv3gfs.sh (hunk 11 above)
Comment (KateFriedman-NOAA) [reply sorc/link_fv3gfs.sh]: Why not commit ecf file changes into the g-w repo? What is this?

**Review by KateFriedman-NOAA — CHANGES_REQUESTED** [review]
I second all of the other developer questions/comments. We need to understand why most of these changes are happening. I see zero incoming changes to ecf files, which is what I expected to see in this PR.

File: parm/config/config.metp (hunk 14 above)
Comment (malloryprow) [reply parm/config/config.metp:20]: @KateFriedman-NOAA With the way things are set up now, the change is fine, as the metp jobs only run for the "gfs" cycle and these settings are for the gfsmetp jobs. However, in the future if the need ever arises that some sort of METplus job needs to be setup and configured to run in the "gdas" cycle, then the if statements would need to go back.

Comment (KateFriedman-NOAA) [reply parm/config/config.metp:20]: Noted, thanks @malloryprow !

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (lgannoaa) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: 1. mpirun is ok to use because host is assigned. It is located in:
     /pe/intel/compilers_and_libraries_2020.4.304/linux/mpi/intel64/bin/mpirun
    Program run without issue see job log:
      /lfs/h2/emc/global/noscrub/Lin.Gan/git/feature-ops-wcoss2/jobs/gfs_emc_vrfy_00.o5065799
2. Using launcher variable in jjob is not within the scope of this work. This job exist in workflow not a new job in wcoss2. Looking at the env file, I do not see every job within workflow are using it. Recommend to handle this request in a separate GitHub issue.

File: jobs/rocoto/arch.sh (hunk 3 above)
Comment (lgannoaa) [reply jobs/rocoto/arch.sh:17]: In wcoss2 source this load_fv3gfs_modules.sh script break things. Please check with changing log, NCO Wei has removed load_fv3gfs_modules.sh on his solution too. If this is needed, maybe it can be modified as machine dependent.

File: jobs/rocoto/earc.sh (hunk 4 above)
Comment (lgannoaa) [reply jobs/rocoto/earc.sh:18]: Same asnwer as before

File: jobs/rocoto/metp.sh (hunk 5 above)
Comment (lgannoaa) [reply jobs/rocoto/metp.sh:19]: The load_fv3gfs_modules.sh conversation is happened in above. It is a repeat conversation here. We can continue this topic in the original message dialog above. Will mark as resolved here.

File: parm/config/config.vrfy (hunk 7 above)
Comment (lgannoaa) [reply parm/config/config.vrfy]: Tracker code manager Jiayi requested the VRFYFSU be turn on.
Will reverse this change to the original.

File: parm/config/config.vrfy (hunk 8 above)
Comment (lgannoaa) [reply parm/config/config.vrfy:36]: It is just temporary. Will fill in when group account installed it. It will not be hardcoded path at the end.

File: parm/config/config.vrfy (hunk 9 above)
Comment (lgannoaa) [reply parm/config/config.vrfy:153]: Similar to the comment above. This is temporary.

File: sorc/checkout.sh (hunk 10 above)
Comment (lgannoaa) [reply sorc/checkout.sh]: Checking out package and link is standard practice in global-workflow. We should do the same for ecf directory. Maybe we can create an official repo for ecflow?

File: sorc/link_fv3gfs.sh (hunk 11 above)
Comment (lgannoaa) [reply sorc/link_fv3gfs.sh]: This is the same question, please communicate in the same communication dialog box above.

File: parm/config/config.vrfy (hunk 8 above)
Comment (aerorahul) [reply parm/config/config.vrfy:36]: The directory `/lfs/h2/emc/global/noscrub/emc.global/git/Fit2Obs` does not exist.

Screen Shot 2022-03-24 at 11 47 07 AM

`VRFYFITS` requires this utility to be installed prior to testing.  
Is it possible that this was built, installed, and tested from your personal space?
In that case, could you please include instructions to build and install to @KateFriedman-NOAA so she can install it in the `save/emc.global/git` space similar to the obsproc packages in the above screen shot.

The same question applies to `ENS_TRACKER`.  Please provide instructions to build and install that utility that typically goes under `HOMEens_tracker`.

File: ush/hpssarch_gen.sh (hunk 12 above)
Comment (lgannoaa) [reply ush/hpssarch_gen.sh:242]: This script wired in the log file using rocoto workflow format. It is modified to include all log file in the log directory, so ecflow log file can also be included in backup set.

File: ush/hpssarch_gen.sh (hunk 13 above)
Comment (lgannoaa) [reply ush/hpssarch_gen.sh:378]: Same solution logic as explained above.

**Comment by lgannoaa** [comment]
> I have pretty much the same questions as Rahul. Also, if code _is_ to be removed, it should just be removed, not commented out.

Historically, many workflow scripts have comments. I prefer to not remove those comments in this scope of work. Maybe  address as an separate issue as needed.

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (aerorahul) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: 1.  Per Carolyn Pasti's presentation on WCOSS2, the correct usage is `mpiexec` and not `mpirun`.
Please see the example from slide 53 from [this](https://www2.emc.ncep.noaa.gov/forms/training/WCOSS2gettingStarted.pdf) presentation.
2. Use of `launcher` is out of the scope of this work.  An issue can be created to resolve this.

File: sorc/checkout.sh (hunk 10 above)
Comment (aerorahul) [reply sorc/checkout.sh]: Storing files in a Private repository is not standard practice anywhere in NOAA-EMC.
Has a need been shown for having a separate set of scripts;  let alone in a users personal account? 
Adding a repository with ecf files is outside of the scope of this PR that enables cycled parallel on WCOSS2 with the verification components.

File: sorc/link_fv3gfs.sh (hunk 11 above)
Comment (aerorahul) [reply sorc/link_fv3gfs.sh]: It is not the same.  
The comment above is a question about introducing a new repository and its associated issues.
Comment here is about moving/linking/duplicating existing scripts and then using them.

And comments are made where they are relevant, even if related.

File: ush/hpssarch_gen.sh (hunk 12 above)
Comment (aerorahul) [reply ush/hpssarch_gen.sh:242]: Does `ecflow` run in development mode for all users?
Does `ecflow` jobs archival jobs call `arch.sh` or `earc.sh`? 
I have not seen those jobs in the `ecf/`.
So how are these being accessed/called?

File: ush/hpssarch_gen.sh (hunk 13 above)
Comment (aerorahul) [reply ush/hpssarch_gen.sh:378]: Same follow-up question as above.

File: parm/config/config.gldas (hunk 16 above)
Comment (lgannoaa) [reply parm/config/config.gldas]: Revert the change.

File: parm/config/config.vrfy (hunk 18 above)
Comment (lgannoaa) [reply parm/config/config.vrfy:104]: This is hardwired. It is removed to allow upper level assignment.

File: parm/config/config.vrfy (hunk 7 above)
Comment (aerorahul) [reply parm/config/config.vrfy]: Ok. And what about he other defaults; e.g. 
```
VSDB_STEP1
VRFYG2OBS
VRFYPRCP
RUNMOS
```
Is there a need for providing external over-rides for these?
```
VRFYRAD
VRFYOZN
VRFYMINMON

```

File: parm/config/config.vrfy (hunk 19 above)
Comment (lgannoaa) [reply parm/config/config.vrfy:144]: Used the same approach as the original. If it needs to be modified, recommend to do in a separate issue ticket.

**Review by aerorahul — COMMENTED** [review]
Please see comments.

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (lgannoaa) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: I do not see problem of using mpirun. The code manager @JiayiPeng-NOAA  should be contacted, if request for change. He is ok with it and checked the output found no issue.

File: jobs/rocoto/arch.sh (hunk 3 above)
Comment (aerorahul) [reply jobs/rocoto/arch.sh:17]: NCO does not use this script in any form whatsoever.  Please correct me if I am incorrect.

File: jobs/JGLOBAL_WAVE_PREP (hunk 2 above)
Comment (lgannoaa) [reply jobs/JGLOBAL_WAVE_PREP]: @JessicaMeixner-NOAA is in control of this JJob, I will take recommendation from her. The reason for this change is the original logic unconditionally make slink to file using global dump directory structure without first checking the availability of the source file.

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (aerorahul) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: Please work w/ whoever it is to bring up to the correct and acceptable usage.  `mpirun` is unacceptable per WCOSS2 instructions.

File: jobs/rocoto/vrfy.sh (hunk 15 above)
Comment (lgannoaa) [reply jobs/rocoto/vrfy.sh:22]: Duplicated comment.

File: jobs/rocoto/vrfy.sh (hunk 6 above)
Comment (lgannoaa) [reply jobs/rocoto/vrfy.sh]: Yes. The $PBS_O_WORKDIR is for PBS. Only pre implementation parallel are running the EMC verification jobs. I will revert this change.

File: parm/config/config.vrfy (hunk 7 above)
Comment (lgannoaa) [reply parm/config/config.vrfy]: The new ecflow run verification job VRFYRAD, VRFYOZN, and VRFYMINMON differently. I adopt the same solution here. Doing so is to test and ensure job run as designed.
I will revert it.

File: parm/config/config.vrfy (hunk 8 above)
Comment (lgannoaa) [reply parm/config/config.vrfy:36]: The path in question here is only a temp place holder.

File: parm/config/config.vrfy (hunk 17 above)
Comment (lgannoaa) [reply parm/config/config.vrfy:54]: If take a look at this file history, it will show this 2 fitdir solution has been used before. Just adopting same solution.

File: sorc/checkout.sh (hunk 10 above)
Comment (lgannoaa) [reply sorc/checkout.sh]: Never said storing file in private repository is a standard practice. Hence the suggestion to create a official repo for ecflow workflow files.
As requested, I will modify this approach to overwrite ecf directory under $HOMEgfs to the new version.

File: sorc/link_fv3gfs.sh (hunk 11 above)
Comment (lgannoaa) [reply sorc/link_fv3gfs.sh]: Agreed to modify the solution above. This is part of the same solution that will be modified.

File: sorc/checkout.sh (hunk 10 above)
Comment (aerorahul) [reply sorc/checkout.sh]: Yes. Please open a separate PR with those changes related internal to `ecf/`
I would like to see the changes in context of NCO's use of them.

File: ush/hpssarch_gen.sh (hunk 12 above)
Comment (lgannoaa) [reply ush/hpssarch_gen.sh:242]: User does not necessary to run ecflow in development mode.
I developed a solution to enable it EMC vrfy and archive jobs. As communicated above, the new ecf directory will replace the original ecf directory in next commit. We can talk about the solution as needed.

File: sorc/checkout.sh (hunk 10 above)
Comment (arunchawla-NOAA) [reply sorc/checkout.sh]: I want to minimize the number of repos that are used to check out a full application from, this is making the global workflow unsustainable. A repo should only really be created when it can stand on its own and be run. Otherwise it should be in one repository only

Comment (lgannoaa) [reply sorc/checkout.sh]: @aerorahul This PR will be the PR for ecflow. I do not understand why it need to have another new PR.

Comment (lgannoaa) [reply sorc/checkout.sh]: @arunchawla-NOAA Thank you for your guidance.

File: ush/hpssarch_gen.sh (hunk 13 above)
Comment (lgannoaa) [reply ush/hpssarch_gen.sh:378]: ok

File: jobs/JGLOBAL_WAVE_PREP (hunk 2 above)
Comment (aerorahul) [reply jobs/JGLOBAL_WAVE_PREP]: @JessicaMeixner-NOAA is unavailable and otherwise occupied with other more pressing issues.
Please highlight specifically what the issue is and how it is impacting WCOSS2 while keeping in mind that this also should work on other machines.
If there is a need for another conditional, please illustrate the need.
If there is a failure encountered with the current logic, please indicate it in a gist for e.g.

File: parm/config/config.vrfy (hunk 19 above)
Comment (aerorahul) [reply parm/config/config.vrfy:144]: When we ported to WCOSS2, as @KateFriedman-NOAA indicates, this needs to be set in the version file and not in the current location.
If unable to do this in this PR, then please open an issue.  It will be assigned appropriately.
It is in the scope of this work.

File: jobs/JGLOBAL_WAVE_PREP (hunk 2 above)
Comment (JessicaMeixner-NOAA) [reply jobs/JGLOBAL_WAVE_PREP]: The availability of the source file is handled in the script, for example If the current day's RTOFS is not available, the script will use the RTOFS from the day before: 
https://github.com/NOAA-EMC/global-workflow/blob/operations/scripts/exgfs_wave_prep.sh#L647-L674 

If that RTOFS file is not there, WW3 prep will error out: 
https://github.com/NOAA-EMC/global-workflow/blob/operations/scripts/exgfs_wave_prep.sh#L711-L721 

I share in the concern expressed by @aerorahul  that the changes you made here will break this for R&D machines and I'm concerned this would also not work on WCOSS2 for EMC or NCO without the commented out sections of the code.

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (arunchawla-NOAA) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: Please use mpiexec, all wcoss2 codes are using that. We have had extensive discussions with NCO on this already

File: jobs/JGLOBAL_WAVE_PREP (hunk 2 above)
Comment (arunchawla-NOAA) [reply jobs/JGLOBAL_WAVE_PREP]: please revert this change. It is already working on wcoss2 and has been extensively tested

File: jobs/rocoto/arch.sh (hunk 3 above)
Comment (arunchawla-NOAA) [reply jobs/rocoto/arch.sh:17]: why are we changing rocoto scripts in this PR? are they not working for rocoto ?

File: ush/hpssarch_gen.sh (hunk 12 above)
Comment (arunchawla-NOAA) [reply ush/hpssarch_gen.sh:242]: perhaps I am missing something but just looking at the logic of the script we will no longer be archiving gdasprep.log, gdasanal.log etc ? Looks like we are only archiving gdaspost*.log ? Or perhaps I am missing the bigger picture ?

File: jobs/JGFS_ATMOS_CYCLONE_TRACKER (hunk 1 above)
Comment (lgannoaa) [reply jobs/JGFS_ATMOS_CYCLONE_TRACKER]: Removed the use of mpirun.

File: jobs/JGLOBAL_WAVE_PREP (hunk 2 above)
Comment (lgannoaa) [reply jobs/JGLOBAL_WAVE_PREP]: Reverted as requested.

File: jobs/rocoto/arch.sh (hunk 3 above)
Comment (lgannoaa) [reply jobs/rocoto/arch.sh:17]: NCO removed scripts source $HOMEgfs/ush/load_fv3gfs_modules.sh due to environment variable been overwrite to default that break ecflow workflow.

File: ush/hpssarch_gen.sh (hunk 12 above)
Comment (lgannoaa) [reply ush/hpssarch_gen.sh:242]: Line 237-242 hardwired to only backup rocoto workflow log file. ecflow workflow does not use the same log file name. Therefore, removed hardwired name and use the line 243 and 378 to take in all log file found in log directory.

**Comment by lgannoaa** [comment]
Close this PR as requested
